### Pass test ------------------------------------------------------------------

# type
expect_silent(check_class("chr", type = "character", fun_name = "fun"))
expect_silent(check_class(2, type = "numeric", fun_name = "fun"))
expect_silent(check_class(-1.4, type = "numeric", fun_name = "fun"))
expect_silent(check_class(2L, type = "integer", fun_name = "fun"))
expect_silent(check_class(TRUE, type = "logical", fun_name = "fun"))
expect_silent(check_class(FALSE, type = "logical", fun_name = "fun"))
expect_silent(check_class(NULL, type = "NULL", fun_name = "fun"))
expect_silent(
  check_class(data.frame(A=c(1,2)), type = "data.frame", fun_name = "fun"))

# allowNULL (Exception from e.g. character)
expect_silent(
  check_class(NULL, type = "character", fun_name = "fun", allowNULL = TRUE))

### Errors ---------------------------------------------------------------------

# var
expect_error(check_class(type = "character", fun_name = "fun"))

# type
expect_error(check_class(var = "character", fun_name = "fun"))
err <- tryCatch(
  check_class(id, type = 12, fun_name = "fun"), 
  error = function(err) err
)
expect_true(
  rlang::inherits_all(err, c("check_class_type_error", "rlang_error")))

# fun_name
expect_error(check_class(var = "chr", type = "type"))
err <- tryCatch(
  check_class(id, type = "numeric", fun_name = 12), 
  error = function(err) err
)
expect_true(
  rlang::inherits_all(err, c("check_class_fun_name_error", "rlang_error")))

# allowNULL
expect_error(
  check_class(id, type = "numeric", fun_name = "fun", allowNULL = "x"))
expect_error(
  check_class(id, type = "numeric", fun_name = "fun", allowNULL = NULL))
expect_error(
  check_class(id, type = "numeric", fun_name = "fun", allowNULL = 12))

err <- tryCatch(
  check_class(id, type = "numeric", fun_name = "fun", allowNULL = "x"), 
  error = function(err) err
)
expect_true(
  rlang::inherits_all(err, c("check_class_allowNULL_error", "rlang_error")))

# individual error generated by function
expect_error(check_class(2, "character", fun_name = "function"))
expect_error(check_class(2, "data.frame", fun_name = "function"))
expect_error(check_class("2", "data.frame", fun_name = "function"))
expect_error(
  check_class(NULL, "character", fun_name = "fun", allowNULL = FALSE))

id <- 1
err <- tryCatch(
  check_class(id, "character", fun_name = "fun", allowNULL = FALSE),
  error = function(err) err
)
expect_true(rlang::inherits_all(err, c("fun_id_error", "rlang_error")))
expect_equal(err$value, 1)
expect_equal(err$current_class, "numeric")

